### PR TITLE
Add nullability check to In-Memory Database

### LINKED ISSUE #10613: Validate requiredness (nullability) in the in-memory database
### Description
I created a model that has a name property which I tried setting as required from both the ModelBuilder and using annotations.
When inspecting the tracked changes it confirms that the property IsNullable == false, out of this 2 scenarios arise:
* When using entity framework in memory, then the change is saved in memory with the field being null.
* When using a connection string for localdb then the following NullReferenceException appears

```
Exception message: Object reference not set to an instance of an object.
Stack trace:       at Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.DatabaseErrorPageMiddleware.System.IObserver<System.Collections.Generic.KeyValuePair<System.String,System.Object>>.OnNext(KeyValuePair`2 keyValuePair)
   at System.Diagnostics.DiagnosticListener.Write(String name, Object value)
   at Microsoft.EntityFrameworkCore.Internal.CoreLoggerExtensions.SaveChangesFailed(IDiagnosticsLogger`1 diagnostics, DbContext context, Exception exception)
   at Microsoft.EntityFrameworkCore.DbContext.SaveChanges(Boolean acceptAllChangesOnSuccess)
   at Microsoft.EntityFrameworkCore.DbContext.SaveChanges()
   at EComm.Front.Data.DbInitializer.Initialize(ApplicationDbContext context) in F:\TFS\Mine\E-Comm\EComm.Front\Data\DbInitializer.cs:line 11
```
This was confirmed through an NUnit test (for the in-memory database) and in an ASP.NET Core 2 application (for the exception version)

### Steps to reproduce
The following code is from the Unit Test

```c#
[TestFixture]
public class DbInitializeTests
{
        [Test]
        public void WhenInitializingTheDb_ItShouldStoreTheItems()
        {
            //arrange
            var dbOptions = new DbContextOptionsBuilder<ApplicationDbContext>();
            dbOptions.UseInMemoryDatabase("testdb");
            var dbContext = new ApplicationDbContext(dbOptions.Options);

            //act
            dbContext.Categories.Add(new Category());

            dbContext.SaveChanges();

            //assert
            Assert.That(dbContext.Categories, Is.Not.Empty);
            Assert.That(dbContext.Categories.First().Name, Is.Not.Null.Or.Empty); // fails here, as it the save went ok up to here.
        }
}

    public class Category : Entity<Guid>
    {
        public string Name { get; set; }

        public string NameInUrlFormat => Name.ToLowerInvariant().Replace(" ", "-");
}

public abstract class Entity<T>
{
     public T Id { get; set; }
}

public class ApplicationDbContext : IdentityDbContext<ApplicationUser>
{
        public ApplicationDbContext(DbContextOptions<ApplicationDbContext> options)
            : base(options)
        {
        }

        public DbSet<Category> Categories { get; set; }

        protected override void OnModelCreating(ModelBuilder builder)
        {
            base.OnModelCreating(builder);
            // Customize the ASP.NET Identity model and override the defaults if needed.
            // For example, you can rename the ASP.NET Identity table names and more.
            // Add your customizations after calling base.OnModelCreating(builder);

            builder.Entity<Category>(
                typeBuilder => 
                    typeBuilder.Property(category => category.Name)
                    .IsRequired()
                    .IsUnicode()
                    .HasMaxLength(100));
        }
 }
```
### Expected Results
The SaveChanges() method should have thrown an exception since the model requirements were not fulfiled

### Further technical details
EF Core version: 2.0.1
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows 10 Pro
IDE: Visual Studio 2017 15.5.2

### Mentions
This issue was also mentioned in the documentation found [here](https://docs.microsoft.com/en-us/ef/core/modeling/required-optional) in the first and last comments.

### PR DESCRIPTION
**Pull request by fagnercarvalho**
Fixes #10613

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/efcore?pullRequest=23094) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/efcore?pullRequest=23094) before we can accept your contribution.<br/><br/>:x: fagnercarvalho [sign now](https://cla.dotnetfoundation.org/dotnet/efcore?pullRequest=23094)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/efcore?pullRequest=23094) it.</sub>

**Review by ajcvickers — APPROVED** [review]
This looks great! Sorry for being so slow to review it.

**Comment by ajcvickers** [comment]
@smitpatel @AndriySvyryd This looks good to me, but I'm not great at noticing places where style, etc. might be wrong. You guys are good at that. :-) Do you want to give it a quick scan?

File: src/EFCore.InMemory/Properties/InMemoryStrings.resx [hunk 1]
```diff
@@ -134,6 +134,9 @@
   <data name="NoQueryStrings" xml:space="preserve">
     <value>There is no query string because the in-memory provider does not use a string-based query language.</value>
   </data>
+  <data name="NullabilityErrorException" xml:space="preserve">
+    <value>Required properties '{requiredProperties}' are missing for instance of entity type '{entityType}' with the key value '{keyValue}'</value>
```
Comment (smitpatel) [inline src/EFCore.InMemory/Properties/InMemoryStrings.resx:138]: may be?
```suggestion
    <value>Required properties '{requiredProperties}' are missing for the instance of entity type '{entityType}' with the key value '{keyValue}'</value>
```

Comment (smitpatel) [reply src/EFCore.InMemory/Properties/InMemoryStrings.resx:138]: Should there be non-sensitive version for this? We are exposing key values.

File: src/EFCore.InMemory/Storage/Internal/InMemoryTable.cs [hunk 2]
```diff
@@ -6,6 +6,7 @@
 using System.Collections.Generic;
 using System.Globalization;
 using System.Linq;
+using System.Runtime.InteropServices;
```
Comment (smitpatel) [inline src/EFCore.InMemory/Storage/Internal/InMemoryTable.cs:9]:
```suggestion
```

File: src/EFCore.InMemory/Storage/Internal/InMemoryTable.cs [hunk 3]
```diff
@@ -188,6 +191,19 @@ public virtual void Create(IUpdateEntry entry)
                 .Select(p => SnapshotValue(p, p.GetKeyValueComparer(), entry))
                 .ToArray();
 
+            var properties = entry.EntityType.GetProperties().ToList();
+            var nullabilityErrors = new List<IProperty>();
+
+            for (var index = 0; index < properties.Count; index++)
```
Comment (smitpatel) [inline src/EFCore.InMemory/Storage/Internal/InMemoryTable.cs:197]: Can combine this for loop with computation of row above.

File: src/EFCore.InMemory/Properties/InMemoryStrings.resx (hunk 1 above)
Comment (ajcvickers) [reply src/EFCore.InMemory/Properties/InMemoryStrings.resx:138]: Good catch! This is why managers should not do code reviews! :-)

File: src/EFCore.InMemory/Storage/Internal/InMemoryTable.cs (hunk 3 above)
Comment (fagnercarvalho) [reply src/EFCore.InMemory/Storage/Internal/InMemoryTable.cs:193]: Makes sense, done.

File: src/EFCore.InMemory/Storage/Internal/InMemoryTable.cs (hunk 2 above)
Comment (fagnercarvalho) [reply src/EFCore.InMemory/Storage/Internal/InMemoryTable.cs]: Done. How can I catch this more easily?

File: src/EFCore.InMemory/Properties/InMemoryStrings.resx (hunk 1 above)
Comment (fagnercarvalho) [reply src/EFCore.InMemory/Properties/InMemoryStrings.resx]: Done @smitpatel. Now to see the entity key values in the required error message you need to enable sensitive logging.

File: src/EFCore.InMemory/Storage/Internal/InMemoryTable.cs [hunk 4]
```diff
@@ -187,26 +186,26 @@ private static List<ValueComparer> GetKeyComparers(IEnumerable<IProperty> proper
...
-                .ToArray();
-
             var properties = entry.EntityType.GetProperties().ToList();
+            var rows = new object[properties.Count];
             var nullabilityErrors = new List<IProperty>();
 
             for (var index = 0; index < properties.Count; index++)
             {
-                HasNullabilityError(properties[index], row[index], nullabilityErrors);
+                var row = SnapshotValue(properties[index], properties[index].GetKeyValueComparer(), entry);
```
Comment (smitpatel) [inline src/EFCore.InMemory/Storage/Internal/InMemoryTable.cs:195]: Consider renaming variables for less confusion
row -> property/column
rows -> row.

File: src/EFCore.InMemory/Storage/Internal/InMemoryTable.cs (hunk 2 above)
Comment (ajcvickers) [reply src/EFCore.InMemory/Storage/Internal/InMemoryTable.cs]: A good IDE should show these as grayed-out or similar so you can see the ones not needed. Also, various code cleanup tools can do this, which we typically run periodically on the codebase anyway.

File: src/EFCore.InMemory/Storage/Internal/InMemoryTable.cs (hunk 4 above)
Comment (fagnercarvalho) [reply src/EFCore.InMemory/Storage/Internal/InMemoryTable.cs]: Sure, done.

**Comment by ajcvickers** [comment]
@fagnercarvalho Thanks for the contribution!

**Comment by devlear** [comment]
I'm not sure if you would like a separate issue created for this, but the changed entity will still be in the ChangeTracker after the DbUpdateException is thrown.

**Comment by AndriySvyryd** [comment]
@devlear That's by design. If any exception happens during `SaveChanges` EF tries to revert to the state before `SaveChanges` was called.

**Pull request merged**